### PR TITLE
feat(msk): IAM access control for clusters

### PR DESCRIPTION
**Pull request by curtiseppel**
MSK now supports IAM access control for client authentication with a MSK
cluster.
https://aws.amazon.com/about-aws/whats-new/2021/05/introducing-iam-access-control-amazon-msk/

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-msk/test/cluster.test.ts [hunk 1]
```diff
@@ -102,18 +102,61 @@ describe('MSK Cluster', () => {
...
-          vpc,
-          encryptionInTransit: {
-            clientBroker: msk.ClientBrokerEncryption.TLS_PLAINTEXT,
-          },
-          clientAuthentication: msk.ClientAuthentication.sasl({
-            scram: true,
-          }),
-        })).toThrow(
-          'To enable SASL/SCRAM authentication, you must only allow TLS-encrypted traffic between clients and brokers.',
+        expect(
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-msk/test/cluster.test.ts:105]: What about a positive test? to make sure the CloudFormation template contains the right config?

**Review by iliapolo — DISMISSED** [review]

File: packages/@aws-cdk/aws-msk/test/cluster.test.ts (hunk 1 above)
Comment (CurtisEppel) [reply packages/@aws-cdk/aws-msk/test/cluster.test.ts:105]: Sorry, I was a bit eager requesting a review. This probably should've been a draft, but I'm getting back to it today.

**Comment by iliapolo** [comment]
@CurtisEppel Can I ask you to resolve the conflicts so the diff is up to date? Thanks!

**Comment by iliapolo** [comment]
@CurtisEppel PR looks good. But build is still failing. Happy to approve after the build passes.

**Review by iliapolo — DISMISSED** [review]
@CurtisEppel PR looks good. But build is still failing. Happy to approve after the build passes.

**Review by iliapolo — APPROVED** [review]

**Comment by gmuslia** [comment]
As far as I can see from the UI the MSK can have both enabled, IAM and SASL/SCRAM, any ideas why we do not allow this in the CDK Level?
Screenshot 2024-05-24 at 15 17 34

**Comment by msambol** [comment]
I was looking at this as well. @curtiseppel, should IAM be independent and moved outside of `SaslAuthProps` ?

https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-msk-alpha/lib/cluster.ts#L367-L372

The auth options can be checked individually. I'm not sure of the history here, it may not always have been this way. 

cc/ @gmuslia

**Comment by msambol** [comment]
@gmuslia Looks like a bug [here](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-msk-alpha/lib/cluster.ts#L689-L711) with the `if/else if`. It's a little weird how CloudFormation grouped [IAM and SCRAM under SASL](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-msk-cluster-sasl.html#cfn-msk-cluster-sasl-iam). I'll see if I can work on a fix. 

@curtiseppel Not sure how involved you are with CDK anymore but jfyi 😄

**Comment by msambol** [comment]
I think this should do it: https://github.com/aws/aws-cdk/pull/31743

cc/ @gmuslia @curtiseppel

**Pull request merged**